### ISSUE TITLE
std::wstring not working properly as custom type for add_option

### ISSUE THREAD

**Issue by boimart1** [henryiii/CLI11#704]
When using CLI11 version 1.9.1, we defined a template specialisation in order to handle using `std::wstring` in `add_option()` calls:
```c++
template <>
bool CLI::detail::lexical_cast<std::wstring>(const std::string& input, std::wstring& output) {
    output = my_utf8_to_utf16(input);
    return true;
}
```
This worked fine in 1.9.1, but no longer compiles in 2.1.2, due to changes in how the underlying `lexical_conversion`/`lexical_cast` system works. Which is fine, there _seems_ to be an alternative: adding `operator<<` and `operator>>` in the CLI namespace, according to [Custom converters](https://cliutils.github.io/CLI11/book/chapters/advanced-topics.html).
```c++
namespace CLI
{
    inline std::istringstream& operator>>(std::istringstream& input, std::wstring& output)
    {
        std::string s;
        input >> s;
        output = my_utf8_to_utf16(s);
        return input;
    }

    inline std::stringstream& operator<<(std::stringstream& output, const std::wstring& input)
    {
        output << my_utf16_to_utf8(input);
        return output;
    }
}  // namespace CLI
```
Unfortunately, this doesn't work either. After digging through the call stack, and rereading the "Custom converters" section, it seems like this sentence is especially relevant:
> These can only be used for "single" size options (so complex, vector, etc. are a separate topic)

because the `lexical_conversion` overload that's being selected is the one which is enabled on `is_mutable_container<AssignTo>::value` (TypeTools.hpp, line 1029) -- that is, `std::wstring` is being treated as a container of `wchar_t`, as opposed to a regular "single object" type.

Evidently, the "1 string to 1 type of some kind" conversion would be preferable (line 1178), so I started trying to work with the internals to tell CLI to treat `wstring` as a regular ol' type:
```c++
    namespace detail
    {
        template <>
        struct classify_object<std::wstring, void>
        {
            static constexpr object_category value{object_category::other};
        };

        template <>
        struct is_mutable_container<std::wstring, void> : public std::false_type
        {
        };
    }  // namespace detail
```
And as expected it seems like the correct `lexical_conversion` is now being called (line 1178), which calls the correct `lexical_assign` (line 1097). However, surprisingly, instead of the correct `lexical_cast` being called (line 1071), we're getting the "assign-from-int" cast (line 1045)... Because as it happens, `std::wstring s; s = 100;` is valid code and so `std::is_assignable<std::wstring&, int>` passes 😔 

At this point, it doesn't seem like there are any more type traits I can safely specialize to get the behavior I want, as I'd rather not try to specialize `std::is_assignable` and possibly break random code.

Basically, have I missed something? Is this a known issue? I'd like to change as little code as possible in the CLI11 2.0+ upgrade, but most of our strings are `wstring`s. I've come up with a workaround using the callback overload of `add_option`, but I'm disappointed to have to work "outside" CLI11's custom type system rather than use it as intended.

**Comment by phlptp** [comment]
I have only on rare occasions worked with wstring so we haven't done much with that type.  

Seems this might be related to the C++ argument dependent lookup(ADL).  std::wstring as a type matches a number of specific types as it seems to be a container and allow for an integer argument so it matches kind of oddly the other operations.  

My suggestions for one more thing to try is 
```cpp
namespace std
{
template <>
bool lexical_cast<wstring>(const std::string& input, wstring& output) {
    output = my_utf8_to_utf16(input);
    return true;
}
 } //namespace std

 namespace CLI::detail
    {
        template <>
        struct classify_object<std::wstring, void>
        {
            static constexpr object_category value{object_category::other};
        };

        template <>
        struct is_mutable_container<std::wstring, void> : public std::false_type
        {
        };
    }  // namespace CLI::detail
```

adding the lexical cast into the std namespace should prioritize it over the ones in CLI detail.  

There might be too many paths with wstring to use the stream operators directly.

**Comment by boimart1** [comment]
Adding a `lexical_cast` template _specialisation_ to `std` would only work if there was a `std::lexical_cast` function template in the first place, but `lexical_cast` belongs to `CLI::detail`. If I were to add something to `std` for the sake of ADL, it would have to either be (1) a new function, or (2) a new function template. In theory we're only allowed to add `std` template specialisations to the `std` namespace, so neither option is permitted by the C++ standard, but I went along with it just to see what would happen.

Adding a function template:
```c++
namespace std
{
    template <typename T, std::enable_if_t<std::is_same_v<T, std::wstring>, int> = 0>
    bool lexical_cast(const std::string& input, T& output)
    {
        output = my_utf8_to_utf16(input);
        return true;
    }
}  // namespace std
```
Gave me the compiler error:
> TypeTools.hpp(1103): error C2668: 'std::lexical_cast': ambiguous call to overloaded function

It sees both the above function template and the function at TypeTools.hpp line 1045 as valid overloads. In this case, they form an overload set where neither is SFINAE'd away, and so it seems like they're both as likely candidates.

Whereas adding a function:
```c++
namespace std
{
    bool lexical_cast(const std::string& input, std::wstring& output)
    {
        output = my_utf8_to_utf16(input);
        return true;
    }
}  // namespace std
```

Actually worked as expected! Well, mostly, there were a few places where we had `add_option<std::optional<std::wstring>, std::wstring>(...)` which passed through `lexical_assign` at line 1147, and the `lexical_cast` at line 1149 is called with an explicit template parameter. This function isn't a template, so it wasn't being used. (I'd suggest removing the explicit `<ConvertTo>` template parameter at line 1149, since few other `lexical_cast` calls in the file provide a template parameter, and it's not necessary in the first place). In any case, it looks like `optional` is handled intrinsically by another `lexical_assign` overload now (as a wrapper type), so just writing `add_option(...)` worked to resolve this issue.

Still, as I said, I'd rather not add stuff to the `std` namespace. So I looked into a way to get this to work by adding the function to `CLI::detail`. Since it's not a function template specialisation, it seems like it needs to be defined before any includes of `CLI/CLI.hpp`, but it otherwise works well.

To summarize, the following works:
```c++
namespace CLI::detail
{
    bool lexical_cast(const std::string& input, std::wstring& output)
    {
        output = my_utf8_to_utf16(input);
        return true;
    }
}  // namespace std

#include <CLI/CLI.hpp>

namespace CLI::detail
{
    template <>
    struct classify_object<std::wstring, void>
    {
        static constexpr object_category value{object_category::other};
    };

    template <>
    struct is_mutable_container<std::wstring, void> : public std::false_type
    {
    };
}  // namespace CLI::detail
```

I'm still unsure why my initial `lexical_cast` template specialisation in `CLI::detail` doesn't compile, so that's something I think I'll look into a little more. ~~I'll also submit a PR to remove the explicit template parameters to some of the `lexical_cast` calls.~~

**Comment by boimart1** [comment]
Ok, so I'm not entirely sure why the ordering is relevant, but writing the following _also_ works, _and_ works for `add_option<T, std::wstring>` calls where `T` isn't a `wstring`.
```c++
#include <CLI/CLI.hpp>

namespace CLI::detail
{
    template <>
    struct classify_object<std::wstring, void>
    {
        static constexpr object_category value{object_category::other};
    };

    template <>
    struct is_mutable_container<std::wstring, void> : public std::false_type
    {
    };

    template <>
    bool lexical_cast(const std::string& input, std::wstring& output)
    {
        output = my_utf8_to_utf16(input);
        return true;
    }
}  // namespace CLI::detail
```
(Given that it's possible to have `lexical_cast` as a template, I don't see the necessity to remove the explicit template parameters to `lexical_cast` calls anymore.)

**Comment by phlptp** [comment]
Glad you got it working!.   I have seen ordering dependent operations in some other situations as well,  Not entirely sure of the exact rule in C++ that is causing it.  Seems to get pretty deep into the overload resolution rules.
